[russelltankaimin] iP

russelltankaimin opened the pull request:
# YodaBot
*A long long time ago, in a galaxy far, far away. Before ascending to the Jedi High Council, Master Yoda was merely a librarian in the Jedi Archives on Coruscant. Growing bored, he built a bot to log every temple-dweller's tasks whilst he seeks adventure outside the temple.*

Get organised with a bot which lets you know what tasks you have! I know it takes a lot of effort but we must try!
> "Do or do not, there is no try" - Yoda <(˶ᵔᵕᵔ˶)>
## Features
- [X] Create a task
- [X] Load existing tasks from database
- [X] Mark, unmark tasks
- [X] Delete tasks
- [X] Search by keyword
- [ ] Timetabling features
## Other fun features
* Responses are yoda-fied 🤣 
## Get started
1. Download from [here](https://github.com/russelltankaimin/ip/releases/tag/A-Jar)
2. Go to terminal and run `java -jar duke.jar`
3. Have fun!
## Additional information
1. If you are a Java programmer and would like to practice Java and Software Engineering, this is the tool for you!
2. You can create your own Java classes and let it interact with existing ones. Entry point of the code is as below:
```java
public static void main(String[] args) {
        new Duke().run();
    }
```
3. Live long and prosper 🖖

Commit:
Add Gradle support

russelltankaimin committed:
(Level-1) Build basic command-response mechanism

russelltankaimin committed:
(Level-2) Implement new TaskList class and new command handling logic

russelltankaimin committed:
(Level-3) Implement new Task class and parsing logic

russelltankaimin committed:
Include abstract Task class and create subclasses of Task

russelltankaimin committed:
(Level-4) Add Support for tracking ToDo, Events and Deadlines

russelltankaimin committed:
Fix chatbot's tone and pre-processing out of bounds bugs

russelltankaimin committed:
Add UI I/O redirection testing cases

russelltankaimin committed:
Create new Exception classes specific to Duke's functionality

russelltankaimin committed:
Make TaskInfoParser more concise with Factory Methods

russelltankaimin committed:
(Level-5) Help Duke handle erroneous commands

russelltankaimin committed:
Create new exception to handle when no such task exists

russelltankaimin committed:
(Level-6) Include deletion of task capability

russelltankaimin committed:
Generalise exceptions to DukeException

russelltankaimin committed:
Convert Task array to Task ArrayList

russelltankaimin committed:
Utilise Enums to handle non-task related commands

russelltankaimin committed:
Add capability to real file for tasks

russelltankaimin committed:
Modify parsing procedures for tasks with exception handling

russelltankaimin committed:
Add file writing and saving capability to Duke

russelltankaimin committed:
(Level-7) Implement save tasks functionality with directory and file creation

russelltankaimin committed:
(Level-8) Integrate use of LocalDate and LocalTime into Duke

russelltankaimin committed:
Add Ui and Storage classes

russelltankaimin committed:
Add different parsers to encapsulate information logic

russelltankaimin committed:
Implement different Command classes

russelltankaimin committed:
Refactor code according to new Ui, Storage, Command classes

russelltankaimin committed:
Repackage classes into packages

russelltankaimin committed:
Rename package names

russelltankaimin committed:
Merge for gradle-support

russelltankaimin committed:
Set up gradle with recommended testing enabled

russelltankaimin committed:
Add JUnit tests for ToDo and TaskInfoParser classes

russelltankaimin committed:
Attempt at writing JavaDocs for some classes

russelltankaimin committed:
Add JavaDocs to all classes in commands package

russelltankaimin committed:
Add JavaDocs to all remaining classes

russelltankaimin committed:
Fix coding style to task classes

russelltankaimin committed:
Fix code to adhere to coding style

russelltankaimin committed:
(Level-9) Add filter capability for Duke to search by keyword

russelltankaimin committed:
Resolve conflicts between JavaDocs branch and master

russelltankaimin committed:
Resolve conflicts between branch-A-CodingStandard with master

russelltankaimin committed:
Set up Duke GUI skeletal code

russelltankaimin committed:
Adapt Command classes and CommandParser to output strings for rendering

russelltankaimin committed:
Update GUI look and mechanics

russelltankaimin committed:
Complete GUI for Duke

russelltankaimin committed:
Update checkstyle from previous level

panavdua commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+import java.io.IOException;
+import java.util.Collections;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import javafx.fxml.FXML;
+import javafx.fxml.FXMLLoader;
+import javafx.geometry.Pos;
+import javafx.scene.Node;
+import javafx.scene.control.Label;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
```
Perhaps, you can separate the import statements of each type with a blank line (as shown in this guide [here](https://se-education.org/guides/conventions/java/intermediate.html))? You can do this in every class wherever you deem suitable!

russelltankaimin replied:
I agree, it would look more organised as well when they are segmented in an appropriate manner. I will take note and change it in the following iteration! Much appreciated!

russelltankaimin replied:
I have tried to add the changes, but apparently checkstyle is not happy with this 😂
<img width="394" alt="Screenshot 2023-02-05 at 7 36 19 AM" src="https://user-images.githubusercontent.com/50812868/216794032-b9d57d03-e44d-4002-9e41-528a8f70a3ff.png">
See here for checkstyle warning
<img width="756" alt="Screenshot 2023-02-05 at 7 36 28 AM" src="https://user-images.githubusercontent.com/50812868/216794054-b336fbdd-cb15-4b51-9f5a-41613ab505ed.png">
I believe that my initial ordering is fine because everything belongs to the same javafx package, then the sub packages are in alphabetical order. I suspect the checkstyle on really complains that the developer add a blank line is when there are 2 or more "master packages" like ```javafx``` and ```javax``` together.

panavdua commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+    /**
+     * Flips the dialog box such that the ImageView is on the left and text on the right.
+     */
+    private void flip() {
```
I really liked the coding quality in this class! But perhaps one thing you can change is add some more clarity in the function name. You can do something like `flipDialogBox()` instead of just `flip()`

russelltankaimin replied:
Yes indeed. I will change that for my next iteration! Thank you!

panavdua commented on src/main/java/parsers/CommandParser.java:
```diff
@@ -0,0 +1,80 @@
+                return new DeleteTaskCommand(commands[1]);
+            case BYE:
+                return new ExitCommand();
+            case TODO:
+            case DEADLINE:
+            case EVENT:
```
In this case, I think you can improve a third-person's understanding of the code by including a comment saying that you have purposedly allowed a fallthrough in the case statements

russelltankaimin replied:
Note taken! I did not think about it that way! Thank you!😊

panavdua reviewed (commented):
Extremely good job with adhering to the coding standards!! Had to search extremely hard to even find a minor problem

russelltankaimin committed:
Add assertions to parsers, file classes, and duke.

The code does not have a lot of assertions but instead throws a lot of
exceptions.

Assertions are a good way to verify assumptions about certain invariants
stemming from the command parsing and their preconditions too.

Let's add assertions to our code more often instead of just relying on
try, catch statements.

russelltankaimin committed:
Improve code quality across files.

Remove redundant code from Duke.

There are many commented-out print statements left in the file after
debugging for Level 10 GUI requirements. There are also many redundant
methods now that Duke runs with javafx.

Code is now shorter and more concise.

Let's try to remove any redundant code left in our codebase.

russelltankaimin committed:
Add streams usage to shorten code

russelltankaimin committed:
Merge pull request #3 from russelltankaimin/branch-A-Streams

Add streams usage to shorten code

russelltankaimin committed:
Merge branch 'master' into branch-A-CodeQuality

russelltankaimin committed:
Merge pull request #2 from russelltankaimin/branch-A-CodeQuality

Improve code quality across files.

russelltankaimin committed:
Merge branch 'master' into branch-A-Assertions

russelltankaimin committed:
Merge pull request #1 from russelltankaimin/branch-A-Assertions

Add assertions to parsers, file classes, and duke.

russelltankaimin committed:
Set up CI

russelltankaimin committed:
Remove yml files first

russelltankaimin committed:
Set up CI with proper yml file

russelltankaimin committed:
Fix checkstyle for tests

russelltankaimin committed:
Add sorting capability for Deadlines

russelltankaimin committed:
Switch target invocation to Launcher instead of Duke

russelltankaimin committed:
Remove scanner

russelltankaimin committed:
Fix header comments for Duke and Main

russelltankaimin committed:
Improve colour scheme and dialog box shapes

russelltankaimin committed:
Improve formatting of long tasks for Event and Deadline

russelltankaimin committed:
Add User Guide and representative screenshot

russelltankaimin committed:
Render image for Linux users

russelltankaimin committed:
Add more personality to Duke
